dali/pipeline/operators/reader/loader/coco_loader.cc

```diff
@@ -12,433 +12,347 @@
-  bool Uint64(uint64_t u) { st_ = kHasNumber; v_.SetUint64(u); return true; }
-  bool Double(double d) { st_ = kHasNumber; v_.SetDouble(d); return true; }
-  bool RawNumber(const char*, SizeType, bool) { return false; }
-  bool String(const char* str, SizeType length, bool) {
-    st_ = kHasString;
-    v_.SetString(str, length);
-    return true;
-  }
-  bool StartObject() { st_ = kEnteringObject; return true; }
-  bool Key(const char* str, SizeType length, bool) {
-    st_ = kHasKey;
-    v_.SetString(str, length);
-    return true;
-  }
-  bool EndObject(SizeType) { st_ = kExitingObject; return true; }
-  bool StartArray() { st_ = kEnteringArray; return true; }
-  bool EndArray(SizeType) { st_ = kExitingArray; return true; }
-
- protected:
-  explicit LookaheadParserHandler(char* str);
-  void ParseNext();
-
- protected:
-  enum LookaheadParsingState {
-    kInit,
-    kError,
-    kHasNull,
-    kHasBool,
-    kHasNumber,
-    kHasString,
-    kHasKey,
-    kEnteringObject,
-    kExitingObject,
-    kEnteringArray,
-    kExitingArray
-  };
-
-  Value v_;
-  LookaheadParsingState st_;
-  Reader r_;
-  InsituStringStream ss_;
-
-  static const int parseFlags = kParseDefaultFlags | kParseInsituFlag;
+#include "dali/pipeline/operators/reader/loader/coco_loader.h"
+#include "dali/pipeline/util/lookahead_parser.h"
+
+namespace dali {
+namespace detail {
+struct ImageInfo {
+  std::string filename_;
+  int original_id_;
+  int width_;
+  int height_;
 };
 
-LookaheadParserHandler::LookaheadParserHandler(char* str) :
-    v_(), st_(kInit), r_(), ss_(str) {
-  r_.IterativeParseInit();
-  ParseNext();
-}
+struct Annotation {
+  int image_id_;
+  int category_id_;
+  std::array<float, 4> box_;
 
-void LookaheadParserHandler::ParseNext() {
-  if (r_.HasParseError()) {
-    st_ = kError;
-    return;
+  void ToLtrb() {
+    box_[2] += box_[0];
+    box_[3] += box_[1];
   }
 
-  r_.IterativeParseNext<parseFlags>(ss_, *this);
-}
-
-class LookaheadParser : protected LookaheadParserHandler {
- public:
-  explicit LookaheadParser(char* str) : LookaheadParserHandler(str) {}
-
-  bool EnterObject();
-  bool EnterArray();
-  const char* NextObjectKey();
-  bool NextArrayValue();
-  int GetInt();
-  double GetDouble();
-  const char* GetString();
-  bool GetBool();
-  void GetNull();
-
-  void SkipObject();
-  void SkipArray();
-  void SkipValue();
-  Value* PeekValue();
-  // returns a rapidjson::Type, or -1 for no value (at end of object/array)
-  int PeekType();
-
-  bool IsValid() { return st_ != kError; }
-
- protected:
-  void SkipOut(int depth);
+  bool IsOver(float min_size_threshold) {
+    return box_[2] >= min_size_threshold && box_[3] >= min_size_threshold;
+  }
 };
 
-bool LookaheadParser::EnterObject() {
-  if (st_ != kEnteringObject) {
-    st_  = kError;
-    return false;
+template<typename T>
+void dump_meta_file(const std::vector<T> &input, const std::string path) {
+  std::ofstream file(path);
```

Comment:
Why not binary output? Don't we loose precision that way?
We could add an option for debug config that would allow human-readable dumping.

Reply:
Done